Mathlib/Tactic/Alias.lean

```diff
@@ -0,0 +1,171 @@
+
+/-- Adds one-way implication declarations, inferring names for them. -/
+syntax (name := aliasLRDots) "alias " ident " ↔ " ".." : command
+
+/-- Elaborates an `alias ←` command. -/
+@[commandElab «alias»] def elabAlias : Elab.Command.CommandElab
+| `(alias $name:ident ← $aliases:ident*) => do
+  let resolved ← resolveGlobalConstNoOverload name
+  let constant ← match (← Lean.getEnv).constants.find? resolved with
+                 | none => throwError "failed to resolve alias LHS"
+                 | some c => pure c
+
+  for a in aliases do
+    let decl ← match constant with
+    | Lean.ConstantInfo.defnInfo d =>
+      pure $ Lean.Declaration.defnDecl {
+        d with name := a.getId
+               value := mkConst resolved
+      }
+    | Lean.ConstantInfo.thmInfo t =>
+      pure $ Lean.Declaration.thmDecl {
+        t with name := a.getId
+               value := mkConst resolved
+      }
+    | _ => throwError "alias only works with def or theorem"
+
+    -- TODO add @alias attribute
+    Lean.addDecl decl
+
+| _ => Lean.Elab.throwUnsupportedSyntax
+
+
+/--
+  Given the type of an iff decl, produce a value for one of the implication
+  directions (determined by `iffmp`).
+-/
+def mkIffMpApp (iffmp: Name) : Expr → (Nat → Expr) → MetaM Expr
+| (Expr.forallE n varType body d), f => do
+     let rest ← mkIffMpApp iffmp body (λ n => mkApp (f (n+1)) (mkBVar n))
+     pure $ mkLambda n d.binderInfo varType rest
+| (Expr.app (Expr.app (Expr.const ``Iff _ _) e1 _) e2 _), f =>
+  pure $ mkApp3 (mkConst iffmp) e1 e2 (f 0)
+| _, _ => throwError "Target theorem must have the form `∀ x y z, a ↔ b`"
+
+/--
+  Given a constant represent an iff decl, adds a decl for one of the implication
+  directions.
+-/
+def aliasIff (ci : ConstantInfo) (al : Name) (isForward : Bool) : MetaM Unit := do
+  let ls := ci.levelParams
+  let t := ci.type
+  let iffmp := if isForward then ``Iff.mp else ``Iff.mpr
+  let v ← mkIffMpApp iffmp t (λ _ => mkConst ci.name (lvls := (ls.map mkLevelParam)))
+  let t' ← Meta.inferType v
+  -- TODO add @alias attribute
+  addDecl $ Lean.Declaration.thmDecl {
+      name := al
+      value := v
+      type := t'
+      levelParams := ls
+  }
+
+/-- Elaborates an `alias ↔` command. -/
+@[commandElab aliasLR] def elabAliasLR : Lean.Elab.Command.CommandElab
+| `(alias $name:ident ↔ $left:binderIdent $right:binderIdent ) => do
+   let resolved ← resolveGlobalConstNoOverload name
+   let constant ← match (← Lean.getEnv).constants.find? resolved with
+                  | none => throwError "failed to resolve alias LHS"
+                  | some c => pure c
+
+   Lean.Elab.Command.liftTermElabM none do
+     match left with
+     | `(binderIdent| $x:ident) =>
+       aliasIff constant x.getId true
+
+     | `(binderIdent| _) => pure ()
+
+     | _ => Lean.Elab.throwUnsupportedSyntax
+
+     match right with
+     | `(binderIdent| $x:ident) =>
+       aliasIff constant x.getId false
```

Comment:
Should we do something about namespaces here?
```lean
namespace A
theorem a : True ↔ True := .rfl
alias a ↔ b c -- should this be A.b?
```

Reply:
I think namespacing already works out okay. I added a test: https://github.com/leanprover-community/mathlib4/pull/293/commits/bd4bd76a53cf7205dd9b220b73778557ed3e8da7

Does that address your concern?

Reply:
Do the `forward` and `backward` lemmas work as well?

Reply:
ah, good point. It appears they don't. I'll take a deeper look...

Reply:
Do you know an easy way to convert `x` into a fully-qualified name?

Reply:
This seems to work: https://github.com/leanprover-community/mathlib4/pull/293/commits/888b4d3e4874f993d3016ced151b0ecb00ad31c6

Reply:
Ah, but the `alias ←` command is still putting everything into the toplevel namespace. Hm...

Reply:
Adjusted that one too: https://github.com/leanprover-community/mathlib4/pull/293/commits/a3376be8ea9863e6c972483db3f461beccec97d5